### ISSUE TITLE
batch training produce poor results

### ISSUE THREAD

**Issue by winstonquock** [deeplearning4j/deeplearning4j#123]
In this multi-layer config, full set training produce reasonable result of F1 but batching the training produce very poor results

```
    trainingSetFetcher.fetch(nSamples)

    val trainingSet = trainingSetFetcher.next()!!
    val col3 = columns(trainingSet.getFeatures(), 7, 8)

    trainingSet.normalizeZeroMeanZeroUnitVariance()
    val layerSizes = array(500, 100, 10) // 500, 100, 10
    val gen = MersenneTwister(123)

    val iteration = 5
    val layerFactory = LayerFactories.getFactory(javaClass<RBM>())
    val listBuilder = NeuralNetConfiguration.Builder()
            .iterations(iteration)!!
            .weightInit(WeightInit.DISTRIBUTION)!!
            .dist(Distributions.normal(gen, 1e-2))!!
            .activationFunction(Activations.tanh())!!
            .hiddenUnit(RBM.HiddenUnit.BINARY)!!
            .visibleUnit(RBM.VisibleUnit.BINARY)!!
            .layerFactory(layerFactory)!!
            .constrainGradientToUnitNorm(true )!!
            .optimizationAlgo(OptimizationAlgorithm.ITERATION_GRADIENT_DESCENT)!!
            .rng(gen)!!
            .learningRate(5e-2.toDouble())!!
            .nIn(trainingSet.numInputs())!!
            .nOut(trainingSet.numOutcomes())!!
            .list(layerSizes.size() + 1)!!
            .hiddenLayerSizes(*layerSizes.toIntArray())

    listBuilder.override({ i, builder ->
                if (i >= layerSizes.size() ) { // 0 based, last one is output
                    builder.layerFactory(DefaultLayerFactory(javaClass<OutputLayer>()))
                    builder.activationFunction(Activations.tanh())
                    builder.lossFunction(LossFunctions.LossFunction.MCXENT)
                    builder.iterations(iteration)
                }
            })
    val conf = listBuilder.build()!!
    val nn = MultiLayerNetwork(conf)
    if (true) { // batched
        var batch = 1
        val iter = ListDataSetIterator(trainingSet.asList(), 500)
        while (iter.hasNext()) {
            println("================== train batch ${batch++} ==================")
            nn.fit(iter.next())
        }
    } else { // full set
        nn.fit(trainingSet)
    }
```
(data set provided separately)

**Comment by hackingbear** [comment]
hi, any discovery on this?

**Comment by agibsonccc** [comment]
I'll be looking into your situation after spark demos are done.

**Comment by agibsonccc** [comment]
So I'm ready to look in to this. How big was your dataset? You're pre processing it right. How are you doing with your test set though? Are you normalizing the training there as well? What batch sizes did you try?

**Comment by agibsonccc** [comment]
Do me a favor: could you try with nn.fit(dataSetIterator) instead?

**Comment by agibsonccc** [comment]
My recommendation here for mini batch learning is going to be using the MultiEpochsIterator with 1 iteration in your configuration. This should help accumulate the error properly across mini batches. Try that and let me know what you get. I would make sure to do varying numbers of epochs as well.

**Comment by winstonquock** [comment]
will try the ```MultiEpochsIterator``` and report back.

**Comment by winstonquock** [comment]
Looks like if I just call
```java
nn.fit(iter)
```
instead of ```nn.fit(iter.next())``` then it would work. ```MultiEpochsIterator``` makes no difference with epoch = 1.

```nn.fit(iter)``` does not do the optimization for each data set iteration but does it after all pretraining. But imaging if we want to update the trained network daily with new data, then we will be effectively doing ```nn.fit(dataSet)``` over and over and the output become trashed. So while the solution proposed work for me to deal with the large initial training set, I feel there is still a problem.

**Comment by agibsonccc** [comment]
There's a bug I'm looking in to with transpose and mmul in nd4j. I'll post here when that's fixed. I'll have some unit tests in place to catch this, shouldn't happen again.

**Comment by agibsonccc** [comment]
Please try again. I uncovered a bug from refactoring in the step function in the line search. This in combination with a matrix multiply bug gave bad feature representations and lack of training. Please let me know how it goes again. I tested this against 3.2 which has a known good optimization setup and it's good now. Thanks!

**Comment by winstonquock** [comment]
I tried the latest nd4j+dl4j. The results are totally trashed even for ```fit(DataSet)```; both the multi epoch run and the full set run produce the same results of predicting only two classes with suspicious outputs
```
predicted
[
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]

  ...
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
  [1.784497350234162E-4, 1.8822799311582914E-4, 1.775640511140682E-4, 1.8349154047422398E-4, 1.7464538474408435E-4]
  [1.7871039995337318E-4, 1.7189318455096213E-4, 1.7932788191501382E-4, 1.751953491692449E-4, 1.8136272054821874E-4]
] // 5599x5
training eval stats: 
Actual Class 0 was predicted with Predicted 1 with count 618 times
Actual Class 0 was predicted with Predicted 4 with count 897 times
Actual Class 1 was predicted with Predicted 1 with count 714 times
Actual Class 1 was predicted with Predicted 4 with count 1079 times
Actual Class 2 was predicted with Predicted 1 with count 170 times
Actual Class 2 was predicted with Predicted 4 with count 231 times
Actual Class 3 was predicted with Predicted 1 with count 377 times
Actual Class 3 was predicted with Predicted 4 with count 518 times
Actual Class 4 was predicted with Predicted 1 with count 421 times
Actual Class 4 was predicted with Predicted 4 with count 574 times
==========================F1 Scores========================================
 0.12945691470096513
===========================================================================
F1: 0.12945691470096513 precision: 0.09688538028651635. recall: 0.19501994075228346. accuracy: 0.4269954143683126
  CLASS[0]: precision: 0.0 recall: 0.0
  CLASS[1]: precision: 0.31043478260869567 recall: 0.39821528165086445
  CLASS[2]: precision: 0.0 recall: 0.0
  CLASS[3]: precision: 0.0 recall: 0.0
  CLASS[4]: precision: 0.17399211882388602 recall: 0.5768844221105528
```
Did you tried my data set?

**Comment by agibsonccc** [comment]
Is this a fresh checkout? Also where is this dataset?

**Comment by winstonquock** [comment]
I sent it to you by private message around the tile I filed the bug. I can sent it again if you need.

**Comment by winstonquock** [comment]
OK, I narrowed it down to the back-prop change even though I did not use the back-prop setting
```
Top of logs
dl4j: commit 825bde4c3e2669d3cab3fc0bfbccb9ecaf07f05e
Author: agibsonccc <0@blix.io>
Date:   Mon Feb 2 19:03:46 2015 -0800

    serializable iteration listeners

nd4j: commit 4ec250b4dffae2c35f7cc452b571c15aa3462c47
Author: agibsonccc <0@blix.io>
Date:   Sun Jan 25 18:51:57 2015 -0800

    change reference to element wise ops.from() to in scope, otherwise multi threading seems to mess with the direct return type of the element wise op

F1: 0.2053452442374114 precision: 0.20370563306319495. recall: 0.20701146379217086. accuracy: 0.4316722917080273
  CLASS[0]: precision: 0.25870646766169153 recall: 0.10297029702970296
  CLASS[1]: precision: 0.3350908025776216 recall: 0.31901840490797545
  CLASS[2]: precision: 0.0752212389380531 recall: 0.04239401496259352
  CLASS[3]: precision: 0.16355140186915887 recall: 0.0782122905027933
  CLASS[4]: precision: 0.1859582542694497 recall: 0.49246231155778897

----------------------------------------------------------------------------------------------------------------------------------------------------------------
dl4j: commit 6c0b09526e3bedf33639e4f4b36bb55b9d29693c
Author: agibsonccc <0@blix.io>
Date:   Thu Feb 5 07:50:03 2015 -0800

     add hooks for backprop, some constant time transpose optimizations

nd4j: commit f8c3e10e5056bc95b17b9fbf9a3747d0df8fdfb4
Author: agibsonccc <0@blix.io>
Date:   Fri Feb 6 11:34:38 2015 -0800

    update


F1: 0.12945691470096513 precision: 0.09688538028651635. recall: 0.19501994075228346. accuracy: 0.4269954143683126
  CLASS[0]: precision: 0.0 recall: 0.0
  CLASS[1]: precision: 0.31043478260869567 recall: 0.39821528165086445
  CLASS[2]: precision: 0.0 recall: 0.0
  CLASS[3]: precision: 0.0 recall: 0.0
  CLASS[4]: precision: 0.17399211882388602 recall: 0.5768844221105528
```

**Comment by winstonquock** [comment]
After reverting the ```IterativeGradientDescent``` change in issue#143, I'm able to do multi-batch training yielding similar scores as full set training. So I think this can be closed.
```
F1: 0.2013885835176132 precision: 0.198924317228046. recall: 0.2039146700939704. accuracy: 0.423030787684926
  CLASS[0]: precision: 0.24508519003931847 recall: 0.12343234323432344
  CLASS[1]: precision: 0.3283289817232376 recall: 0.28053541550474065
  CLASS[2]: precision: 0.06896551724137931 recall: 0.0399002493765586
  CLASS[3]: precision: 0.16527196652719664 recall: 0.08826815642458101
  CLASS[4]: precision: 0.18696993060909792 recall: 0.48743718592964824
```

**Comment by agibsonccc** [comment]
So are there any changes to be made, or do you think you could submit a pull request? Thanks!

**Comment by winstonquock** [comment]
See my comment in https://github.com/SkymindIO/deeplearning4j/issues/143 apparently I get it work by reverting that ```IterativeGradientDescent``` class there, but I don't know what's right fix since you must have a reason to make that change in the first place.

**Comment by agibsonccc** [comment]
Fixed with your latest pull request.
